### PR TITLE
Bound malformed tokens, indices and placement recursion

### LINKED ISSUE #2150: [Bug]: Malformed tokens, indices and placement references are not bounded safely
### What happened?

The supplied five malformed inputs contain an unterminated comment, an unterminated string, an invalid triangle index, a cyclic placement and a missing placement reference. Processing must terminate within the test timeout without native out-of-bounds access or a crash. Existing paths can scan beyond the file or follow invalid references. These are robustness cases, not claims that malformed files comply with IFC.

### Reproduction

[minimal IFC input](https://github.com/rihokirss/engine_web-ifc/blob/215706d3d9414ac2c80f11c91d4b289c114ae36a/tests/functional/fixtures/malformed-input/cyclic-placement.ifc); [all fixtures](https://github.com/rihokirss/engine_web-ifc/tree/215706d3d9414ac2c80f11c91d4b289c114ae36a/tests/functional/fixtures/malformed-input); [MalformedInput.spec.js](https://github.com/rihokirss/engine_web-ifc/blob/215706d3d9414ac2c80f11c91d4b289c114ae36a/tests/functional/MalformedInput.spec.js).

The regression test is added by the proposed PR and is **not yet in upstream main**. The linked IFC can be downloaded and opened with an upstream build independently. To run the test, use a clean checkout of the PR branch (with Emscripten configured):

```sh
git clone --single-branch --branch codex/pr-malformed-input https://github.com/rihokirss/engine_web-ifc.git web-ifc-repro-malformed-input
cd web-ifc-repro-malformed-input
npm ci
npm run build-release
npm test -- --runTestsByPath tests/functional/MalformedInput.spec.js
```

Against the upstream build: **3 failed, 2 passed, 5 total**. Against the proposed fix: **5 passed**. The test is included on the linked branch; to compare builds, `WEB_IFC_API` can point to another built `web-ifc-api-node.js`.

### Version

Source `4760418f6bc2751f49692e7cab9e40abb3038cfd` (0.0.78), freshly built Node WASM; Node 24.13.0, Emscripten 4.0.10.

**Proposed fix:** https://github.com/ThatOpen/engine_web-ifc/pull/2151.

### PR DESCRIPTION
**Pull request by rihokirss**
Fixes #2150.

Bound EOF scanning and triangle access, detect cyclic/deep placements, and guard missing FILE_SCHEMA. Tokenizer failures now propagate through the loader to both OpenModel entry points, which close the failed model and return -1 instead of exposing a partial model. Bounded child-process tests cover malformed tokens after valid headers and records, callback/buffer loading, chunk boundaries and successful loads after repeated failures. Text-escape decoding is handled separately.

Validation on this branch's fresh Emscripten 4.0.10 builds: **11 PR regression tests and 89 existing functional/API tests pass in Node 24.13.0; 6 tests pass in headless Chromium using the browser WASM.** The six browser cases cover late tokenizer rejection; native ASan/UBSan also passed 2400 tokenizer cases.

The test file and fixtures are introduced by this PR and are not yet in upstream main. The test constructs additional edge cases from the fixtures. To run it, use a clean checkout of the PR branch (with Emscripten configured):

```sh
git clone --single-branch --branch codex/pr-malformed-input https://github.com/rihokirss/engine_web-ifc.git web-ifc-repro-malformed-input
cd web-ifc-repro-malformed-input
npm ci
npm run build-release
npm test -- --runTestsByPath tests/functional/MalformedInput.spec.js
```

Reproduction: [minimal IFC input](https://github.com/rihokirss/engine_web-ifc/blob/a40ac991d92ee878cb67f4a45e59e6286abdf79a/tests/functional/fixtures/malformed-input/cyclic-placement.ifc); [all fixtures](https://github.com/rihokirss/engine_web-ifc/tree/a40ac991d92ee878cb67f4a45e59e6286abdf79a/tests/functional/fixtures/malformed-input); [MalformedInput.spec.js](https://github.com/rihokirss/engine_web-ifc/blob/a40ac991d92ee878cb67f4a45e59e6286abdf79a/tests/functional/MalformedInput.spec.js).

Build environment: Node 24.13.0, Emscripten 4.0.10.

### CODE REVIEW

File: src/cpp/web-ifc/geometry/IfcGeometryLoader.cpp [hunk 1]
```diff
@@ -3848,6 +3848,18 @@ namespace webifc::geometry
 
   glm::dmat4 IfcGeometryLoader::GetLocalPlacement(uint32_t expressID, glm::dvec3 vector) const
   {
+    if (_activePlacements.size() >= 512 || !_activePlacements.insert(expressID).second)
+    {
+      spdlog::error("[GetLocalPlacement()] Cyclic or excessively deep placement {}", expressID);
+      return glm::dmat4(1);
+    }
+    struct PlacementScope
```
Comment (beachtom) [inline src/cpp/web-ifc/geometry/IfcGeometryLoader.cpp:3926]: What is this?

Comment (rihokirss) [reply src/cpp/web-ifc/geometry/IfcGeometryLoader.cpp:3926]: This is a local RAII cleanup guard for cycle detection. Once a placement ID is inserted into _activePlacements, the destructor removes it whenever this call exits, including early returns and exceptions. This keeps the set limited to the current recursion path, so reusing a placement later is not mistaken for a cycle.
The cyclic-placement.ifc fixture contains a self-reference (#22=IFCLOCALPLACEMENT(#22,#21)), which would otherwise recurse indefinitely.

Comment (beachtom) [reply src/cpp/web-ifc/geometry/IfcGeometryLoader.cpp:3926]: No the struct - I was not sure why you are using a struct here and where else it is being used?

Comment (rihokirss) [reply src/cpp/web-ifc/geometry/IfcGeometryLoader.cpp:3926]: It isn't used anywhere else — both the type and the scope instance are local to GetLocalPlacement. I used a struct specifically to give that instance a destructor: when the call exits, it removes the placement ID automatically, including on early returns or exceptions. This avoids repeating erase() at every exit point. The name could be clearer; ActivePlacementGuard with a short comment would express its purpose better

**Review by beachtom — APPROVED** [review]

**Comment by rihokirss** [comment]
Thanks for updating the branch. I noticed the merge resolution dropped the missing/invalid FILE_SCHEMA guards in both OpenModel and OpenModelFromCallback. We should retain those checks alongside ALLOW_INCOMPATIBLE_SCHEMA_ALIASES, so a missing header returns -1 and closes the model instead of throwing. I can push a small follow-up to preserve both.

**Comment by beachtom** [comment]
Yeah I spotted that - will sort
